subprojects/docs/src/docs/userguide/developing-plugins/writing_in_kotlin.adoc

```diff
@@ -0,0 +1,329 @@
+
+This implicit link between affects the Gradle versions your plugin supports. It's easy to build a plugin using one version of Gradle that that has functions or classes that are not present in other versions of Gradle.
+
+This might result in runtime exceptions, such as `NoSuchMethodError`, if a plugin uses a functionality of the Gradle API that is missing in the Gradle being used to run the plugin. This might happen the plugin is run by an older Gradle version that is missing a function or class, _or_ if the plugin uses a function or class that has been removed in newer versions (which can happen when using deprecated or internal Gradle features).
+
+Supporting different versions of Gradle can be tested using https://docs.gradle.org/current/userguide/test_kit.html#sub:gradle-runner-gradle-version[TestKit].
+
+[[sec:redistributed_gradle_api]]
+=== Redistributed Gradle API
+
+To avoid the implicit link on a Gradle version that comes with using `gradleApi()`, an alternative is to use the https://docs.nokee.dev/manual/gradle-plugin-development.html[Nokee redistributed Gradle API artifacts]. These are like `gradleApi()` except they allow specifying a specific version of the Gradle API.
+
+[subs="attributes"]
+[source,kotlin]
+----
+dependencies {
+    compileOnly("dev.gradleplugins:gradle-api:{gradleVersion}")
+}
+----
+
+TIP: Because it is provided by Gradle at runtime, `gradle-api` must be a `compileOnly` dependency.
+
+This makes the version of the Gradle API being used to build the plugin more explicit, removing the ambiguity avoids the implicit link between `gradleApi()` and the build version of Gradle.
+
+On the other hand, you cannot use newer APIs if your plugin is running on a recent version of Gradle.
+
+[[sec:gradle_kotlin_dsl]]
+=== gradleKotlinDsl()
+
+[source,kotlin]
+----
+dependencies {
+    compileOnly(gradleKotlinDsl())
+}
+----
+
+The `gradleKotlinDsl()` contains helpers to work better with Kotlin.
+
+For an example:
+
+* `tasks.withType<Jar> {}` (instead of `tasks.withType(Jar::class.java) {}`)
+* `val commonMain by getting {}` (instead of `getByName("commonMain") {}`)
+* `"implementation"("group:artifact:version")` (instead of `add("implementation", "group:artifact:version")`)
+* Even `gradleKotlinDsl()` is provided by `gradleKotlinDsl()` !
+
+You can use optionally it to build your plugins, but it's not mandatory.
+
+It is part of the Gradle distribution. You will find it under `~/.gradle/wrapper/dists/gradle-$version-bin/$sha/gradle-$version/lib/gradle-kotlin-dsl-$version.jar`.
+
+When writing `build.gradle.kts`, `gradleKotlinDsl()` is in the classpath and provide helpers to work better with Kotlin.
+
+TIP: Because it is provided by Gradle at runtime, `gradleKotlinDsl()` must be a `compileOnly` dependency.
+
+[[sec:kotlin_stdlib]]
+=== kotlin-stdlib
+
+[source,kotlin]
+----
+dependencies {
+    compileOnly("org.jetbrains.kotlin:kotlin-stdlib:$version")
```

Comment:
🥇 To help with cases like this, the [Toolbox project offers a Gradle Runtime Compatibility class to map this knowledge](https://github.com/gradle-plugins/toolbox/blob/main/subprojects/gradle-plugin-development/src/main/java/dev/gradleplugins/GradleRuntimeCompatibility.java#L175). It offers other API as well like the [minimum supported Java version per Gradle version](https://github.com/gradle-plugins/toolbox/blob/main/subprojects/gradle-plugin-development/src/main/java/dev/gradleplugins/GradleRuntimeCompatibility.java#L132) and, soon, the [latest patch version of a minor version](https://github.com/gradle-plugins/toolbox/blob/main/subprojects/gradle-plugin-development/src/main/java/dev/gradleplugins/GradleRuntimeCompatibility.java#L294) as well as the [latest patched minor version for a major version](https://github.com/gradle-plugins/toolbox/blob/main/subprojects/gradle-plugin-development/src/main/java/dev/gradleplugins/GradleRuntimeCompatibility.java#L351).